src/gc.c

```diff
@@ -4206,6 +4206,21 @@ JL_DLLEXPORT jl_value_t *jl_gc_internal_obj_base_ptr(void *p)
     return NULL;
 }
 
+JL_DLLEXPORT jl_value_t *jl_gc_internal_obj_base_ptr(void *p)
+{
+    uint8_t restore = gc_all_tls_states == NULL;
+    if (__unlikely(restore)) {
+        gc_n_threads = jl_atomic_load_acquire(&jl_n_threads);
+        gc_all_tls_states = jl_atomic_load_relaxed(&jl_all_tls_states);
+    }
```

Comment:
I am pretty sure it would be forbidden to access this data except when GC is running (and thus the values are already set). The `gc_all_tls_states` info for heap.norm_pools seems unlikely to be safe to access

Reply:
@fingolfin I agree with Jameson here, `jl_gc_internal_obj_base_ptr` is only used for the tests right? All other uses should be inside the GC?

Reply:
We essentially call `jl_gc_internal_obj_base_ptr` in two ways:
- in our mark functions (and functions they call)
- in our stack scanning code, so by the GC callbacks installed via `jl_gc_set_cb_root_scanner` resp. `jl_gc_set_cb_task_scanner`. 

All of these should only be run during GC. Or at least that's how they were set up originally, I have not checked recently.

Reply:
OK, so I see, the code in `test/gcext/LocalTest.jl` invokes `jl_gc_internal_obj_base_ptr` in other places -- that's perhaps a bug in the testing harness then... Looking at it, I guess @rbehrends did it this way to ensure the task scanning is *really* called and stressed. So I guess those tests should be adjusted.

But our actual use case in GAP.jl does not suffer from this.